src/System.Private.ServiceModel/src/System/ServiceModel/Channels/MessageContent.cs

```diff
@@ -21,23 +22,21 @@ internal abstract class MessageContent : HttpContent
         protected MessageEncoder _messageEncoder;
         protected Stream _stream = null;
         private bool _disposed;
+        protected TaskCompletionSource<bool> _writeCompletedTcs;
 
         public MessageContent(Message message, MessageEncoder messageEncoder)
         {
             _message = message;
             _messageEncoder = messageEncoder;
+            _writeCompletedTcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
 
             SetContentType(_messageEncoder.ContentType);
             PrepareContentHeaders();
         }
 
-        public Message Message
-        {
-            get
-            {
-                return _message;
-            }
-        }
+        public Message Message { get { return _message; } }
+
+        internal  Task WriteCompletionTask { get { return _writeCompletedTcs?.Task; } }
```

Comment:
Nit: none of the other dereferences of _writeCompletedTcs check for null.  Is this one strictly needed, or should the others check too?